components/esp_dns/esp_dns.c

```diff
@@ -0,0 +1,318 @@
+    int ret = -1;
+    switch (config->protocol) {
+    case DNS_PROTOCOL_UDP:
+        ret = init_udp_dns(handle);
+        break;
+
+    case DNS_PROTOCOL_TCP:
+        ret = init_tcp_dns(handle);
+        break;
+
+    case DNS_PROTOCOL_DOT:
+        ret = init_dot_dns(handle);
+        break;
+
+    case DNS_PROTOCOL_DOH:
+        ret = init_doh_dns(handle);
+        break;
+
+    default:
+        ESP_LOGE(TAG, "Unsupported DNS protocol: %d", config->protocol);
+        ret = -1;
+        break;
+    }
+
+    if (ret != 0) {
+        ESP_LOGE(TAG, "Failed to initialize DNS protocol %d: Error Code: %d",
+                 config->protocol, handle->last_error);
+
+        /* Clean up resources */
+        if (handle->lock != NULL) {
+            vSemaphoreDelete(handle->lock);
+        }
+        free(handle);
+        xSemaphoreGive(g_dns_global_mutex);
+        return NULL;
+    }
+
+    /* Set global handle */
+    g_dns_handle = handle;
+    handle->initialized = true;
+
+    /* Release global mutex */
+    xSemaphoreGive(g_dns_global_mutex);
+
+    ESP_LOGI(TAG, "DNS module initialized successfully with protocol %d", config->protocol);
+    return handle;
+}
+
+/**
+ * Cleanup and release resources associated with a DNS module handle.
+ *
+ * @param handle DNS module handle previously obtained from esp_dns_init()
+ * @return 0 on success, non-zero error code on failure
+ */
+int esp_dns_cleanup(esp_dns_handle_t handle)
+{
+    /* Validate parameters */
+    if (handle == NULL) {
+        ESP_LOGE(TAG, "Invalid handle (NULL)");
+        return -1;
+    }
+
+    /* Take the handle mutex */
+    if (xSemaphoreTake(handle->lock, portMAX_DELAY) != pdTRUE) {
+        ESP_LOGE(TAG, "Failed to take handle mutex during cleanup");
+        return -1;
+    }
+
+    /* Clean up protocol-specific resources */
+    int ret = 0;
+    switch (handle->config.protocol) {
+    case DNS_PROTOCOL_UDP:
+        ret = cleanup_udp_dns(handle);
+        break;
+
+    case DNS_PROTOCOL_TCP:
+        ret = cleanup_tcp_dns(handle);
+        break;
+
+    case DNS_PROTOCOL_DOT:
+        ret = cleanup_dot_dns(handle);
+        break;
+
+    case DNS_PROTOCOL_DOH:
+        ret = cleanup_doh_dns(handle);
+        break;
+
+    default:
+        ESP_LOGW(TAG, "Unknown protocol during cleanup: %d", handle->config.protocol);
+        ret = -1;
+        break;
+    }
+
+    /* Mark as uninitialized */
+    handle->initialized = false;
+
+    /* Release and delete mutex */
+    xSemaphoreGive(handle->lock);
+    vSemaphoreDelete(handle->lock);
+
+    /* Take global mutex before modifying global handle */
+    if (g_dns_global_mutex != NULL && xSemaphoreTake(g_dns_global_mutex, portMAX_DELAY) == pdTRUE) {
+        /* Clear global handle if it matches this one */
+        if (g_dns_handle == handle) {
+            g_dns_handle = NULL;
+        }
+
+        xSemaphoreGive(g_dns_global_mutex);
+    }
+
+    /* Empty the handle */
+    memset(handle, 0, sizeof(handle));
```

Comment:
Shouldn't the handle be freed?

Reply:
Check and cleanup

Reply:
The buffers in the handle are managed by each transport type and may not need freeing. I'll add the cleanup logic in individual cleanup function.